src/Model/AuditManager.php

```diff
@@ -33,17 +35,37 @@ class AuditManager implements AuditManagerInterface
     protected $readers = [];
 
     /**
-     * @var ContainerInterface
+     * @var SymfonyContainerInterface
      */
     protected $container;
 
-    public function __construct(ContainerInterface $container)
+    /**
+     * @var ContainerInterface|null
+     */
+    private $psrContainer;
+
+    // NEXT_MAJOR: Remove SymfonyContainerInterface parameter and only use ContainerInterface parameter
+    public function __construct(SymfonyContainerInterface $container, ?ContainerInterface $psrContainer = null)
     {
         $this->container = $container;
+        $this->psrContainer = $psrContainer;
     }
 
     public function setReader($serviceId, array $classes)
     {
+        // NEXT_MAJOR: Remove this "if" block.
+        if (null !== $this->psrContainer && !$this->psrContainer->has($serviceId)) {
+            @trigger_error(sprintf(
+                'Not registering the audit reader "%s" with tag "%s" is deprecated since'
+                .' sonata-project/admin-bundle 3.x and will not work in 4.0.'
+                .' You MUST add "%s" tag to the service "%s".',
+                $serviceId,
+                AddAuditReadersCompilerPass::AUDIT_READER_TAG,
+                AddAuditReadersCompilerPass::AUDIT_READER_TAG,
+                $serviceId
```

Comment:
```suggestion
                'Not registering the audit reader "%1$s" with tag "%2$s" is deprecated since'
                .' sonata-project/admin-bundle 3.x and will not work in 4.0.'
                .' You MUST add "%2$s" tag to the service "%1$s".',
                $serviceId,
                AddAuditReadersCompilerPass::AUDIT_READER_TAG
```